lib/src/cancelable_operation.dart

```diff
@@ -150,6 +150,65 @@ class CancelableOperation<T> {
       {FutureOr<R> Function(Object, StackTrace)? onError,
       FutureOr<R> Function()? onCancel,
       bool propagateCancel = false}) {
+    if (onError != null && onCancel != null) {
+      return thenOperation<R>((value, completer) {
+        completer.complete(onValue(value));
+      }, onError: (error, stack, completer) {
+        completer.complete(onError(error, stack));
+      }, onCancel: (completer) {
+        completer.complete(onCancel());
+      }, propagateCancel: propagateCancel);
+    } else if (onError != null) {
+      return thenOperation<R>((value, completer) {
+        completer.complete(onValue(value));
+      }, onError: (error, stack, completer) {
+        completer.complete(onError(error, stack));
+      }, propagateCancel: propagateCancel);
+    } else if (onCancel != null) {
+      return thenOperation<R>((value, completer) {
+        completer.complete(onValue(value));
+      }, onCancel: (completer) {
+        completer.complete(onCancel());
+      }, propagateCancel: propagateCancel);
+    } else {
+      return thenOperation<R>((value, completer) {
+        completer.complete(onValue(value));
+      }, propagateCancel: propagateCancel);
+    }
+  }
+
+  /// Creates a new cancelable operation to be completed
+  /// when this operation completes or is cancelled.
```

Comment:
This is a long first paragraph (guildeline is one line).
Not that I can find a good shorter version, though.